### ISSUE TITLE
Support for o1-mini and o1-preview

### ISSUE THREAD

**Issue by jame25** [hibobmaster/matrix_chatgpt_bot#56]
Since there is now API access to these models, could you please add support for them. I've tried generating a new API key, however I still get the following error:  `state=finished raised KeyError`

**Comment by hibobmaster** [comment]
Doc: https://platform.openai.com/docs/guides/reasoning

![图片](https://github.com/user-attachments/assets/17340518-ddd5-4cc8-a523-7c7bb8b0f87f)

There is no compatible issue i think. Do you have the [required tier](https://help.openai.com/en/articles/9824962-openai-o1-preview-and-o1-mini-usage-limits-on-chatgpt-and-the-api#h_0944b89caf) to use this model?

**Comment by jame25** [comment]
Ah, that is probably the issue. I am only tier 1. I have however recently become able to interact with the API via the playground:

![Capture](https://github.com/user-attachments/assets/6cecf73a-f844-46ed-9e8a-f4456a5807bd)

Thanks for taking the time to respond.

**Comment by jame25** [comment]
FYI, I got an email an hour ago confirming that I now have API access to both models (o1-mini and o1-preview). Also the related webpage has been updated:

![Capture](https://github.com/user-attachments/assets/ef1493f8-22ef-458e-8c8b-37df5f6c166c)

I am still facing the same error:  `state=finished raised KeyError`

**Comment by hibobmaster** [comment]
Can you share me your config with sensitive information redacted.

Besides, what's in container log?

**Comment by jame25** [comment]
The container log:

> 2024-11-20 12:15:07,453 - INFO - matrix chatgpt bot start.....
> 2024-11-20 12:15:10,841 - INFO - Successfully login via password
> Mismatch in keys payload of device MatrixChatGPTBot (xxxxx) of user @bot:matrix.org (@bot:matrix.org).
> 2024-11-20 12:15:25,930 - INFO - Message received in room XXXX
> User | !gpt hello
> 2024-11-20 12:15:30,729 - gpt - ERROR - RetryError[<Future at 0x7fb025e2f790 state=finished raised KeyError>]
> 2024-11-20 12:15:30,729 - ERROR - RetryError[<Future at 0x7fb025e2f790 state=finished raised KeyError>]
> 2024-11-20 12:15:32,003 - INFO - Message received in room XXXX
> bot | > <@user:matrix.org> !gpt hello
> Something went wrong, please try again or contact admin.

My config:

> HOMESERVER="https://matrix.org"
> USER_ID="@bot:matrix.org"
> PASSWORD=xxxxx
> DEVICE_ID="MatrixChatGPTBot"
> ROOM_ID="!xxxxx:matrix.org"
> OPENAI_API_KEY="xxxxxxxxxxxxxx"
> GPT_API_ENDPOINT="https://api.openai.com/v1/chat/completions"
> GPT_MODEL="o1-preview"
> MAX_TOKENS=4000
> TOP_P=1.0
> PRESENCE_PENALTY=0.0
> FREQUENCY_PENALTY=0.0
> REPLY_COUNT=1
> SYSTEM_PROMPT="You are ChatGPT, a large language model trained by OpenAI. Respond conversationally"
> TEMPERATURE=0.8
> TIMEOUT=120.0

**Comment by hibobmaster** [comment]
Does `gpt-3.5-turbo` work on well? 

Can you use curl to do some test as well? If curl work on well, please paste the output here with sensitive info redacted.
https://platform.openai.com/docs/api-reference/chat/create?lang=curl

**Comment by jame25** [comment]
Yes, other models including `gpt-3.5-turbo` work fine. The curl request for `o1-preview` returned the following: 

`BadRequestError: Error code: 400 - {'error': {'message': "Unsupported value: 'messages[0].role' does not support 'system' with this model.", 'type': 'invalid_request_error', 'param': 'messages[0].role', 'code': 'unsupported_value'}}`

Which in turn, lead me to [stackoverflow](https://stackoverflow.com/questions/78981559/why-the-new-openai-model-o1-preview-doesnt-support-system-role). And subsequently the creation of the following python script - which does work with the o1-mini and o1-preview models:

```
import os
import requests
import json

# Retrieve your OpenAI API key from an environment variable
api_key = os.getenv("OPENAI_API_KEY")

# Check if the API key is available
if not api_key:
    raise ValueError("Please set the OPENAI_API_KEY environment variable.")

# Set up the URL and headers for the request
url = "https://api.openai.com/v1/chat/completions"

headers = {
    "Content-Type": "application/json",
    "Authorization": f"Bearer {api_key}"
}

# Define your prompt
prompt = "Can you tell me a joke?"

# Structure the messages accordingly
messages = [
    {
        "role": "user",
        "content": prompt  # Use the prompt directly as a string
    }
]

# Define the data payload for the request
data = {
    "model": "o1-preview",
    "messages": messages
}

# Send the POST request
response = requests.post(url, headers=headers, json=data)

# Check if the request was successful
if response.status_code == 200:
    # Extract the assistant's reply content
    reply_content = response.json()["choices"][0]["message"]["content"]

    # Print the assistant's reply (since it's already a string)
    print("Response from GPT:", reply_content)
else:
    # Print the error
    print(f"Status Code: {response.status_code}")
    print("Response:")
    print(response.json())
```

**Comment by hibobmaster** [comment]
Try: `ghcr.io/hibobmaster/matrixchatgptbot:sha-c2d6ecfaf62f539f9e5ac0d321aa6c56002ec437`

**Comment by jame25** [comment]
That version works great with both o1 models. Thank you.

**Comment by jame25** [comment]
I've just gotten access to `o1` and `o3-mini` and it looks like `/src/gptbot.py` may need updating. Currently neither of these models are supported.

**Comment by hibobmaster** [comment]
Try: `ghcr.io/hibobmaster/matrixchatgptbot:v1.9.0`

**Comment by jame25** [comment]
That version works great. Thank you for the quick fix.

**Comment by jame25** [comment]
Sorry for asking, but could you provide an update for `o3` support? Perhaps implement a wildcard so new models are compatible or drop validation altogether and let OpenAI API return an error if the model doesn’t exist?

>      import fnmatch
>
>     ALLOWED_PATTERNS = ["o*-mini", "o*", "gpt-*"]  # as broad as you like
>     if not any(fnmatch.fnmatch(model_name, pat) for pat in ALLOWED_PATTERNS):
>         raise ValueError(f"{model_name} doesn’t match any allowed model pattern")
